### ISSUE TITLE
Can't create any transactions on develop branch at b0a8522d9

### ISSUE THREAD

**Issue by nmarley** [dashpay/dash#4699]
### Describe the issue

I am not able to create any transaction it seems, after compiling and running dashd on the `develop` branch at commit b0a8522d9. My wallet is unlocked and has plenty of tDash.

If I try and prepare a proposal collateral (gobject prepare) transaction, I get a message:

```
% ./src/dash-cli -testnet gobject prepare 0 1 441763201 7b22656e645f65706f6368223a313634353731373832342c226e616d65223a226e6d7465737432222c227061796d656e745f61646472657373223a227950796170344e6d374c536647325a457a6975514678784253456f556d566a545747222c227061796d656e745f616d6f756e74223a372c2273746172745f65706f6368223a313634353534353032342c2274797065223a312c2275726c223a2268747470733a2f2f6e6d74657374322e74657374227d
error code: -32603
error message:
Error making collateral transaction for governance object. Please check your wallet balance and make sure your wallet is unlocked.

% ./src/dash-cli --version -testnet
Dash Core RPC client version v0.17.0.3-b0a8522d9
```

If I try and send to a fresh address in my own wallet, I get this message from the command line:

```
% ./src/dash-cli -testnet sendtoaddress ybnko7ptc3qycBsZySQJR6fxSThVcvC8Yc 0.00000001
error code: -4
error message:
Error: This transaction requires a transaction fee of at least 1231455.75809168
% ./src/dash-cli -testnet sendtoaddress ybnko7ptc3qycBsZySQJR6fxSThVcvC8Yc 1
error code: -4
error message:
Error: This transaction requires a transaction fee of at least 1231455.76345744
% ./src/dash-cli -testnet sendtoaddress ybnko7ptc3qycBsZySQJR6fxSThVcvC8Yc 7
error code: -4
error message:
Error: This transaction requires a transaction fee of at least 1231455.7688232
```

If I try and send via the GUI, I get these messages (please see attached images).

<img width="1077" alt="dashcore-send-issue1" src="https://user-images.githubusercontent.com/1735354/155328000-6246c16c-f1c6-44b7-b309-b5f31c8a007c.png">

<img width="1087" alt="dashcore-send-issue2" src="https://user-images.githubusercontent.com/1735354/155328016-c456d156-526c-4139-a12c-24875f3a4d2b.png">

#### If so, please list the steps to reproduce below:
1. git checkout b0a8522d9
2. cd depends && make CC=clang CXX=clang++
3. cd .. && bash autogen.sh && ./configure --prefix=$PWD/depends/x86_64-apple-darwin19.6.0 CC=clang CXX=clang++ && make
4. ./dash/src/qt/dash-qt -testnet&
5. Then do any of the dash-cli commands above, or just try and send some tDash somewhere via the GUI.

### Expected behaviour
It should send the Dash without the refillkeypool error.

### Actual behaviour
An error is thrown about being unable to create a transaction and says to refillkeypool (doing this does not help, still gives the same error).

### Machine specs:
- OS: MacOS Catalina v10.15.7
- CPU: 2.4 GHz 8-Core Intel Core i9
- RAM: 64 GB 2667 MHz DDR4
- Disk size: Free 505.41 GB / Capacity 1 TB
- Disk Type (HD/SDD): SSD

### Any extra information that might be useful in the debugging process.

Relevant debug.log entries after `gobject prepare`:

```
2022-02-23T13:21:05Z gobject_prepare -- params: 0 1 441763201 7b22656e645f65706f6368223a313634353731373832342c226e616d65223a226e6d7465737432222c227061796d656e745f61646472657373223a227950796170344e6d374c536647325a457a6975514678784253456f556d566a545747222c227061796d656e745f616d6f756e74223a372c2273746172745f65706f6368223a313634353534353032342c2274797065223a312c2275726c223a2268747470733a2f2f6e6d74657374322e74657374227d, data: {"end_epoch":1645717824,"name":"nmtest2","payment_address":"yPyap4Nm7LSfG2ZEziuQFxxBSEoUmVjTWG","payment_amount":7,"start_epoch":1645545024,"type":1,"url":"https://nmtest2.test"}, hash: 60ff45061aa21e28ba2db63cbaab28a00fa9bc4b8b73f9006f0284314e13ea7f
2022-02-23T13:21:05Z [default wallet] CWallet::GetBudgetSystemCollateralTX -- Error: Keypool ran out, please call keypoolrefill first
```

Wallet has plenty of funds and not locked

```
% ./src/dash-cli -testnet getwalletinfo
{
  "walletname": "",
  "walletversion": 120200,
  "balance": 200194.99999765,
  "coinjoin_balance": 0.00000000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 0.00000000,
  "txcount": 2,
  "timefirstkey": 1628013916,
  "keypoololdest": 1628013916,
  "keypoolsize": 999,
  "keys_left": 999,
  "paytxfee": 0.00000000,
  "scanning": false,
  "private_keys_enabled": true
}
```

**Comment by thephez** [comment]
Interesting. Just checked on my machine and it's working fine for me (same commit, different OS).

**Comment by nmarley** [comment]
Very interesting. I'll try and git bisect it to find the commit where this started, but it will take a while because I have to recompile dependencies and everything from scratch each time because of linker issues. Maybe a functional test as @PastaPastaPasta suggested would help reduce the debug time.

**Comment by nmarley** [comment]
Running `keypoolrefill` doesn't help with this specific wallet, but I haven't tried creating a new one. Will try that and see what happens. The wallet I am currently using was created on v0.17.0.3 I believe.

**Comment by nmarley** [comment]
After running a git bisect to debug this, it looks like 254e12221779fcfb9ac35e7aa549b39ba788786d is the change which introduced this.

I created a patch on develop which reverts d25c0f21b and 254e12221 and this seems to work to fix the issue, which you can see at https://github.com/nmarley/dash/commit/df0cc17cd440507ccdcd2cc6232ce82c84867f2d.

I am assuming this is because the wallet is not an HD wallet but need to research a bit more. Will add some debug around here and see what comes up. I imagine that a bug was introduced with the different wallet flags.

**Comment by nmarley** [comment]
Ok, I've created and funded 2 brand-new wallets from the branch above https://github.com/nmarley/dash/commit/df0cc17cd440507ccdcd2cc6232ce82c84867f2d, 1 of them is an HD wallet and the other is not.

I re-started DashCore at b0a8522d9d1a072f24810ca32b06b21e1b0fbc1c with each wallet loaded in turn, and the HD wallet had no issue but the regular wallet was giving the same error as the above. So it seems like the changes introduced some bug for regular non-HD wallets.

 I will test this on Linux later just to verify it's not an OS thing, but I'm pretty sure I've got this nailed down to HD vs non-HD.
